[export] Add missing builtin ops.

zhxchen17 opened the pull request:
Summary: Fixing issue https://github.com/pytorch/pytorch/issues/113778

Test Plan: eyes.

Differential Revision: D51436177




cc @avikchaudhuri @gmagogsfm @tugsbayasgalan @angelayi @suo @ydwu4

ydwu4 reviewed (approved):
Looks good to me! non-blocking: do you mind copying over the test in https://github.com/pytorch/pytorch/issues/113778 so that we could guard against any accidental regressions for built-in ops?

zhxchen17 committed:
[export] Add missing builtin ops. (#113982)

Summary:

Fixing issue https://github.com/pytorch/pytorch/issues/113778

Test Plan: eyes.

Reviewed By: ydwu4

Differential Revision: D51436177

zhxchen17 commented:
@pytorchbot merge

titaiwangms commented on torch/_export/verifier.py:
```diff
@@ -101,6 +102,11 @@ def allowed_builtin_ops(self) -> List:
+            operator.neg,
+            operator.abs,
+            math.ceil,
+            math.floor,
```
Weirdly, math.floor and math.ceil still not work.

titaiwangms replied:
https://hud.pytorch.org/pr/114762

```
2023-11-29T16:27:56.4907052Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/interpreter.py", line 195, in run_node
2023-11-29T16:27:56.4907250Z     return getattr(self, n.op)(n.target, args, kwargs)
2023-11-29T16:27:56.4907881Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/pass_base.py", line 208, in call_function
2023-11-29T16:27:56.4908166Z     raise ExportPassBaseError(f"Unsupported target type: {target}")
2023-11-29T16:27:56.4908674Z torch._export.pass_base.ExportPassBaseError: Unsupported target type: <built-in function ceil>
2023-11-29T16:27:56.4908681Z 
2023-11-29T16:27:56.4909205Z While executing %ceil : [num_users=1] = call_function[target=math.ceil](args = (%_local_scalar_dense_14,), kwargs = {})
2023-11-29T16:27:56.4909323Z Original traceback:
2023-11-29T16:27:56.4909587Z   File "onnx/test_fx_to_onnx_with_onnxruntime.py", line 155, in func
2023-11-29T16:27:56.4909740Z     torch.tensor([math.ceil(x.item())]),
```

zhxchen17 replied:
@titaiwangms oh I think this is a different error from the verifier one raised here. It's from dynamic shape constraints by looking at the stack:
```
2023-11-29T16:27:56.4895828Z _ TestFxToOnnxWithOnnxRuntime_op_level_debug_True_dynamic_shapes_False_model_type_TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM.test_sympy_operatons_return_numeric _
2023-11-29T16:27:56.4897782Z /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/__init__.py:685: UserWarning: torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, please use torch.set_default_dtype() and torch.set_default_device() as alternatives. (Triggered internally at /var/lib/jenkins/workspace/torch/csrc/tensor/python_tensor.cpp:453.)
2023-11-29T16:27:56.4897908Z   _C._set_default_tensor_type(t)
2023-11-29T16:27:56.4898947Z /opt/conda/envs/py_3.8/lib/python3.8/site-packages/numba/core/types/__init__.py:108: FutureWarning: In the future `np.long` will be defined as the corresponding NumPy scalar.
2023-11-29T16:27:56.4899072Z   long_ = _make_signed(np.long)
2023-11-29T16:27:56.4900136Z /opt/conda/envs/py_3.8/lib/python3.8/site-packages/scipy/__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.2)
2023-11-29T16:27:56.4900614Z   warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "
2023-11-29T16:27:56.4900743Z Traceback (most recent call last):
2023-11-29T16:27:56.4901436Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/passes/infra/pass_manager.py", line 270, in __call__
2023-11-29T16:27:56.4901553Z     res = fn(module)
2023-11-29T16:27:56.4902211Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/passes/infra/pass_base.py", line 40, in __call__
2023-11-29T16:27:56.4902347Z     res = self.call(graph_module)
2023-11-29T16:27:56.4903329Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py", line 140, in call
2023-11-29T16:27:56.4903460Z     val = super().call(graph_module)
2023-11-29T16:27:56.4904049Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/pass_base.py", line 427, in call
2023-11-29T16:27:56.4904276Z     result = self.call_submodule(graph_module, tuple(inputs))
2023-11-29T16:27:56.4904905Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/pass_base.py", line 390, in call_submodule
2023-11-29T16:27:56.4905046Z     interpreter.run(*inputs_data)
2023-11-29T16:27:56.4905597Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/interpreter.py", line 138, in run
2023-11-29T16:27:56.4905730Z     self.env[node] = self.run_node(node)
2023-11-29T16:27:56.4906339Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/pass_base.py", line 231, in run_node
2023-11-29T16:27:56.4906458Z     return super().run_node(n)
2023-11-29T16:27:56.4907052Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/interpreter.py", line 195, in run_node
2023-11-29T16:27:56.4907250Z     return getattr(self, n.op)(n.target, args, kwargs)
2023-11-29T16:27:56.4907881Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_export/pass_base.py", line 208, in call_function
2023-11-29T16:27:56.4908166Z     raise ExportPassBaseError(f"Unsupported target type: {target}")
2023-11-29T16:27:56.4908674Z torch._export.pass_base.ExportPassBaseError: Unsupported target type: <built-in function ceil>
2023-11-29T16:27:56.4908681Z 
```
Thanks for letting me know and feel free to raise a separate issue if you want.

@tugsbayasgalan is it hard to rewrite assertion pass without ExportPassBase?
